src/types.rs

```diff
@@ -956,12 +957,79 @@ impl Types {
+                let mut maybe_constr: HashSet<Ident> = HashSet::new();
+                let mut to_be_checked: Vec<&Types> = vec![body];
+                while let Some(tys) = to_be_checked.pop() {
+                    match &tys.types {
+                        TypeF::Arrow(s, t) => {
+                            to_be_checked.push(s);
+                            to_be_checked.push(t);
+                        }
+                        TypeF::Forall {
+                            var: var_,
+                            body: body_,
+                            ..
+                        } => {
+                            // check for shadow (forall x. forall x. foo)
+                            if var_ != var {
+                                to_be_checked.push(body_);
+                            }
+                        }
+                        TypeF::Enum(erows) if var_kind == VarKind::EnumRows => {
+                            maybe_constr.clear();
+                            for eitem in erows.iter() {
+                                match eitem {
+                                    EnumRowsIteratorItem::Row(row) => {
+                                        maybe_constr.insert(*row);
+                                    }
+                                    EnumRowsIteratorItem::TailVar(var_) => {
+                                        if var_ == var {
+                                            constr.extend(&maybe_constr);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        TypeF::Record(rrows) if var_kind == VarKind::RecordRows => {
+                            maybe_constr.clear();
+                            for ritem in rrows.iter() {
+                                match ritem {
+                                    RecordRowsIteratorItem::Row(row) => {
+                                        to_be_checked.push(row.types);
+                                        maybe_constr.insert(row.id);
+                                    }
+                                    RecordRowsIteratorItem::TailDyn => (),
+                                    RecordRowsIteratorItem::TailVar(var_) => {
+                                        if var_ == var {
+                                            constr.extend(&maybe_constr);
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        _ => (),
+                    }
+                }
+
+                let constr_ncl: RichTerm = Term::Array(
+                    Array::from_iter(
+                        constr
+                            .into_iter()
+                            .map(|id| Term::Str(id.label().to_owned().into()).into()),
```

Comment:
I think there is a `From<Label>` instance for `NickelString`

```suggestion
                            .map(|id| Term::Str(id.into()).into()),
```